### ISSUE TITLE
Support ASP.NET precompile and build WebDeploy package

### ISSUE THREAD

**Issue by shibayan** [microsoft/dotnet-framework-docker#157]
`microsoft/dotnet-framework:4.7.2-sdk` image is not included ASP.NET precompile and packaging tools.

I want to install WebBuildTools in the image. Ref #152

### Results

```
Microsoft (R) Build Engine version 15.7.179.6572 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  WebApplication12 -> C:\codebuild\tmp\src786232751\src\github.com\shibayan\codebuild-aspnet\WebApplication12\bin\WebApplication12.dll
  Transformed Web.config using C:\codebuild\tmp\src786232751\src\github.com\shibayan\codebuild-aspnet\WebApplication12\Web.Release.config into obj\Release\TransformWebConfig\transformed\Web.config.
  Copying all files to temporary location below for package/publish:
  obj\Release\AspnetCompileMerge\Source.
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\Web\Transform\Microsoft.Web.Publishing.AspNetCompileMerge.targets(132,5): error : Can't find the valid AspnetMergePath [C:\codebuild\tmp\src786232751\src\github.com\shibayan\codebuild-aspnet\WebApplication12\WebApplication12.csproj]
```

```
Microsoft (R) Build Engine version 15.7.179.6572 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  WebApplication12 -> C:\codebuild\tmp\src159701527\src\github.com\shibayan\codebuild-aspnet\WebApplication12\bin\WebApplication12.dll
  Transformed Web.config using C:\codebuild\tmp\src159701527\src\github.com\shibayan\codebuild-aspnet\WebApplication12\Web.Release.config into obj\Release\TransformWebConfig\transformed\Web.config.
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\Web\Deploy\Microsoft.Web.Publishing.MSDeploy.Common.targets(55,5): error : Package/Publish task Microsoft.Web.Publishing.Tasks.CreateProviderList failed to load Web Deploy assemblies. Microsoft Web Deploy is not correctly installed on this machine. Microsoft Web Deploy v3 or higher is recommended. [C:\codebuild\tmp\src159701527\src\github.com\shibayan\codebuild-aspnet\WebApplication12\WebApplication12.csproj]
```

**Comment by LouisMT** [comment]
I've tried adding `Microsoft.VisualStudio.Workload.WebBuildTools` to the `sdk` image to compare the image size with and without the `WebBuildTools`, but NGen is returning an error after adding this package (`ngen update` says something about an incorrect version (32/64 bit) is being used) which I don't have time to look into at the moment.

I would be really useful if this could be included in the image, maybe in a separate `sdk-web` image like how it's done for .NET Core?

**Comment by MichaelSimons** [comment]
@HongGit - Did you cover scenarios like this in your scenario testing?  What is your recommendation?

**Comment by HongGit** [comment]
No, I did not cover this scenario in our testing.  

@MichaelSimons aspnet to support precompile, the aspnet_merge.exe is part of the .NET SDK tools.  I see that Targeting Packs are installed, is there a reason that the .NET SDK tools aren't included?

**Comment by shibayan** [comment]
Precompiling a view is very important when creating a Docker Image of an ASP.NET application, so I'd like it to work with it.

**Comment by me-mapspeople** [comment]
I am currently seeing:

`GenerateMsdeployManifestFiles:
  Generate source manifest file for Web Deploy package/publish ...
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\VisualStudio\v15.0\Web\Microsoft.Web.Publishing.targets(3758,5): error : Package/Publish task Microsoft.Web.Publishing.Tasks.IsCleanMSDeployPackageNeeded failed to load Web Deploy assemblies. Microsoft Web Deploy is not correctly installed on this machine. Microsoft Web Deploy v3 or higher is recommended.`

when trying to use `msbuild.exe` with `/p:DeployOnBuild=true` and `/p:PublishProfile=XXX` in the `microsoft/dotnet-framework:4.7.2-sdk-windowsservercore-1803` image.

I presume this is related. Adding above details to help people find this issue.

However, does anyone have a proper solution for this?

**Comment by shibayan** [comment]
I found workaround this issue. The problem occurs because `SDK40ToolsPath` always points to 4.6.1 SDK.

You can workaround this by overriding the `SDK40ToolsPath` at csproj.

```xml
<SDK40ToolsPath>C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.7.2 Tools\</SDK40ToolsPath>
```

It will not be necessary for MSBuild, which is included with Visual Studio 2019.
ref #214

**Comment by kernelcoredump** [comment]
Even with the latest _4.7.2-windowsservercore-ltsc2019_ image containing the VS2019 Build Tools, I was still getting the originally cited error in https://github.com/Microsoft/dotnet-framework-docker/issues/157#issue-326809284 from the failure in _CreateProviderList_.

Fixed by installing Web Deploy 3.6 from https://www.iis.net/downloads/microsoft/web-deploy.  I now just use a builder image from a Dockerfile like:
```
FROM mcr.microsoft.com/dotnet/framework/sdk:4.7.2-windowsservercore-ltsc2019
RUN powershell -Command \
  New-Item -ItemType directory -Path C:\install ;\
  Invoke-WebRequest https://download.microsoft.com/download/0/1/D/01DC28EA-638C-4A22-A57B-4CEF97755C6C/WebDeploy_amd64_en-US.msi -OutFile C:\install\WebDeploy_amd64_en-US.msi ;\
  Start-Process -FilePath msiexec.exe -ArgumentList '/i','C:\install\WebDeploy_amd64_en-US.msi','/passive','/quiet','/qn' -NoNewWindow -Wait
```

Perhaps the Web Deploy package/files (less than 10MB) should be added as part of the SDK image.

**Comment by MichaelSimons** [comment]
Assigning to Matt to investigate the feasibility of including web deploy.

**Comment by mthalman** [comment]
There are two separate issues here:
1. ASP.NET precompile not working. The fix for this issue will be handled by #214.
2. Make WebDeploy available in the SDK image.  This current issue will focus on solving this aspect.
